source/MaterialXGraphEditor/Graph.cpp

```diff
@@ -4029,6 +4031,106 @@ void Graph::drawGraph(ImVec2 mousePos)
+            {
+
+                isNodeGraph |= _currUiNode->getNode()->getImplementation()->isA<mx::NodeGraph>();
+            
+                for (ed::NodeId selected : selectedNodes)
+                {
+                    int pos = findNode((int) selected.Get());
+                    if (pos >= 0)
+                    {
+                        UiNodePtr node = _graphNodes[pos];
+                        
+                        if (node->getNode() != nullptr)
+                        {
+                            isNodeGraph |= node->getNode()->getImplementation()->isA<mx::NodeGraph>();
+                        }
+                        
+                        _copiedNodes.insert(
+                                std::pair<UiNodePtr, UiNodePtr>(_graphNodes[pos], 
+                                                                nullptr));                                      
+                    }
+                }
+
+            }
+
+            // delete nodes we just copied
+
+            if (!_copiedNodes.empty() && !isNodeGraph)
+
+            {
+            
+                for (std::map<UiNodePtr, UiNodePtr>::iterator iter = _copiedNodes.begin(); iter != _copiedNodes.end(); iter++)
+                {
+                    UiNodePtr node = iter->first;
+                    deleteNode(node);
+                }
+        
+                // create subgraph
+            
+                UiNodePtr nodegraphnodeptr = addNode("nodegraph", "", "");
+                            
+                _currUiNode = nodegraphnodeptr;
+
+                ed::SetNodePosition(_currUiNode->getId(), mousePos);                
+
+                // dive inside
+            
+                if (_currUiNode->getNodeGraph() != nullptr)
+                {
+                
+                    savePosition();
+                    _graphStack.push(_graphNodes);
+                    _pinStack.push(_currPins);
+                    _sizeStack.push(_graphTotalSize);
+                    mx::NodeGraphPtr implGraph = _currUiNode->getNodeGraph();
+                    _initial = true;
+                    _graphNodes.clear();
+                    _isNodeGraph = true;
+                    setRenderMaterial(_currUiNode);
+                    ed::DeselectNode(_currUiNode->getId());
+                    _currUiNode = nullptr;
+                    _currGraphElem = implGraph;
+                    _currGraphName.push_back(implGraph->getName());
+                    buildUiNodeGraph(implGraph);
+                    ed::NavigateToContent();
+                }
+            
+                // paste                                
+            
+                for (std::map<UiNodePtr, UiNodePtr>::iterator iter = _copiedNodes.begin(); iter != _copiedNodes.end(); iter++)
+                {
+                    copyUiNode(iter->first);
+                }
+                _addNewNode = true;
+            }
+
+        }
+                        
```

Comment:
A small thing here. I think you want to clear `_copiedNodes` if not performing the copy to make the action fail in an atomic manner. 

In fact maybe you want to clear it all the time since your just using copy+paste as an internal implementation convenience.

Reply:
hey ! Thanks ! 
I should able to make those changes this weekend ! :)

Reply:
I added the call to clear the copied nodes. I also cleaned up a sneaky #include iostream that I had accidently left in there..